### ISSUE TITLE
[Bug]: Release of 5.73.0 leading to change in required IAM permission

### ISSUE THREAD

**Issue by kentedstrom** [hashicorp/terraform-provider-aws#40104]
### Terraform Core Version

1.9.8

### AWS Provider Version

5.73.0

### Affected Resource(s)

Affected Resource(s):
**RDS Database**: The issue occurred during the attempt to provision an RDS instance, although it affected all infrastructure provisioning tasks.
**EC2 (DescribeTags)**: The error specifically referenced the ec2:DescribeTags action, which was blocked due to insufficient IAM permissions. This action is part of the EC2 API and is required during infrastructure provisioning, especially when describing resources and tags.
IAM Role/Policy: The IAM role we utilize was impacted by changes in required permissions following the new release of the Terraform AWS provider. This role and its perms have not been changed and were working fine before 5.73.0

### Expected Behavior

**Successful Infrastructure Provisioning:** Terraform should have successfully provisioned all infrastructure, including the RDS database, and any other resources defined in the configuration, without encountering authorization errors.

**Appropriate IAM Permissions:** The IAM role used by Terraform should have had the necessary permissions to perform all required actions, including ec2:DescribeTags and any other related permissions needed for the provisioning of resources across environments. **This role has not been changed and we tested reverting back to 5.72.1 without the EC2(DescribeTags) permission and it worked as intended.**

**Backward Compatibility:** The Terraform AWS provider should have been backward compatible, meaning that the new version (5.73.0) should not have introduced any breaking changes or new IAM permission requirements that would block deployments using previously valid configurations.

### Actual Behavior

Provisioning Failed Due to Permission Errors: The infrastructure provisioning failed across all environments when trying to create resources, including the RDS database. The error message specifically indicated that the IAM role assumed by Terraform which was not authorized to perform the ec2:DescribeTags action due to missing identity-based policy permissions. **This permission was not required before this version.**

We verified the issue was with 5.73.0 by reverting back to 5.72.1 without the ec2:DescribeTags perm and it worked as intended.

### Relevant Error/Panic Output Snippet

```shell
#1:
Error: listing tags for EC2 (Elastic Compute Cloud) <thing> (subnet-foobar): operation error EC2: DescribeTags, https response error StatusCode: 403, RequestID: 813ab14b-56d7-4b4b-958c-47a9dce4f1ef, api error UnauthorizedOperation: You are not authorized to perform this operation. User: arn:aws:sts::foobar:assumed-role/foobar/foobar is not authorized to perform: ec2:DescribeTags because no identity-based policy allows the ec2:DescribeTags action
with module.subnets.data.aws_subnet.all["subnet-foobar],
on .terraform/modules//main.tf line 32, in data "aws_subnet" "all":
32: data "aws_subnet" "all" {
Error: listing tags for EC2 (Elastic Compute Cloud) <thing> (subnet-foobar): operation error EC2: DescribeTags, https response error StatusCode: 403, RequestID: 69607a85-5728-459a-9a64-c6cc920aee85, api error UnauthorizedOperation: You are not authorized to perform this operation. User: arn:aws:sts::foobar:assumed-role/foobar/aws-go-sdk-foobar is not authorized to perform: ec2:DescribeTags because no identity-based policy allows the ec2:DescribeTags action
with module.subnets.data.aws_subnet.all["subnet-foobar"],
on .terraform/modules/subnets/main.tf line 32, in data "aws_subnet" "all":
32: data "aws_subnet" "all" {
Error: listing tags for EC2 (Elastic Compute Cloud) <thing> (subnet-foobar): operation error EC2: DescribeTags, https response error StatusCode: 403, RequestID: b4847c91-999d-4f4f-ac71-9e015da9f602, api error UnauthorizedOperation: You are not authorized to perform this operation. User: arn:aws:sts::foobar:assumed-role/foobar/aws-go-sdk-foobar is not authorized to perform: ec2:DescribeTags because no identity-based policy allows the ec2:DescribeTags action
with module.subnets.data.aws_subnet.all["subnet-foobar"],
on .terraform/modules/subnets/main.tf line 32, in data "aws_subnet" "all":
32: data "aws_subnet" "all" {
```

### Terraform Configuration Files

```terraform
data "aws_availability_zones" "available" {
  state = "available"
}

# All data subnets for this VPC and Cluster
data "aws_subnets" "subnets" {
  filter {
    name   = "vpc-id"
    values = [var.vpc_id]
  }

  tags = {
    "riot.io/resource-type/data" = true
    "riot.io/cluster-name"       = local.cluster_name
  }
}

# All data subnets for this VPC and Cluster with detailed information such as available IPs
data "aws_subnet" "all" {
  for_each = toset(data.aws_subnets.subnets.ids)
  id       = each.key
}
```

### Steps to Reproduce

1. Ensure Terraform is Installed

2.  Use an IAM Role Without Necessary Permissions: Set up an IAM role with the following permissions (or lack thereof):

  The role should have limited permissions (specifically missing ec2:DescribeTags and other necessary permissions for resource provisioning).
  This IAM role should be assumed by Terraform during infrastructure provisioning.

3. Create or Use Existing Terraform Configuration: Create a Terraform configuration that provisions resources across different AWS services (e.g., EC2, RDS, etc.). Include resources that rely on ec2:DescribeTags or similar EC2 API actions.

4. Configure Terraform to Use the AWS Provider: Set up the AWS provider configuration in Terraform, ensuring that you’re using a version of the AWS provider prior to the update (or ensure Terraform automatically pulls the latest version).

5. Run Terraform Plan/Apply: Execute the terraform apply or terraform plan command, allowing Terraform to provision resources using the IAM role from step 2.

6. Observe the Error: Upon running the terraform apply or terraform plan command, observe the error message similar to:

```
UnauthorizedOperation: You are not authorized to perform this operation. 
User: arn:aws:sts::777474137401:assumed-role/blah is not authorized to perform: 
ec2:DescribeTags because no identity-based policy allows the ec2:DescribeTags action.
Confirm Permission Issue: Verify that the IAM role being used does not have the ec2:DescribeTags permission or other permissions introduced by the new AWS provider version (5.73.0 or higher), which prevents successful resource provisioning.
```

### Important Factoids

1. We are utilizing an in-house IAC application but are confident that it's not a contributing factor
2. We tested reverting to version 5.72.1 and that fixed the issue even WITHOUT the additional IAM permission of `ec2:describetags`

### References

We suspect this commit may be a cause:
https://github.com/hashicorp/terraform-provider-aws/commit/802b6329bf0e12f2d9bfd4180141c2598a09c84e 

### Would you like to implement a fix?

No

**Comment by novekm** [comment]
@justinretzolk can you take a look into this when you have some time?

**Comment by jar-b** [comment]
Hello and thanks for your issue report! 

First, I'd like to start with some background on the change which caused this, introduced with #39761. (The commit referenced in the original issue, while also related to transparent tagging, was released in `v5.74.0` and would not have impacted any of the Plugin SDK-based data sources used in the provided configuration). 

Starting in `v5.0.0` the provider introduced the concept of ["transparent tagging"](https://hashicorp.github.io/terraform-provider-aws/resource-tagging/#transparent-tagging) - a mechanism which allows us to centralize tag handling code across all resources. Transparent tagging wraps the tagging APIs available for a given service to manipulate resource tags, usually `TagResource`, `UntagResource`, and `ListTagsForResource`. For EC2 the last API is actually named [`DescribeTags`](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeTags.html).

While the initial focus was on resources, we've introduced similar patterns for data sources over the course of subsequent `v5.x.x` minor releases. In `v5.73.0`, the `aws_subnet` data source was migrated to transparent tagging with the following change.


```diff
git diff  v5.72.1..v5.73.0 internal/service/ec2/vpc_subnet_data_source.go
diff --git a/internal/service/ec2/vpc_subnet_data_source.go b/internal/service/ec2/vpc_subnet_data_source.go
index 65c1dffa75..c64bfa2b3c 100644
--- a/internal/service/ec2/vpc_subnet_data_source.go
+++ b/internal/service/ec2/vpc_subnet_data_source.go
@@ -20,6 +20,8 @@ import (
 )

 // @SDKDataSource("aws_subnet")
+// @Tags(identifierAttribute="id")
+// @Testing(generator=false)
 func dataSourceSubnet() *schema.Resource {
        return &schema.Resource{
                ReadWithoutTimeout: dataSourceSubnetRead,
@@ -138,7 +140,6 @@ func dataSourceSubnet() *schema.Resource {
 func dataSourceSubnetRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
        var diags diag.Diagnostics
        conn := meta.(*conns.AWSClient).EC2Client(ctx)
-       ignoreTagsConfig := meta.(*conns.AWSClient).IgnoreTagsConfig

        input := &ec2.DescribeSubnetsInput{}

@@ -235,10 +236,6 @@ func dataSourceSubnetRead(ctx context.Context, d *schema.ResourceData, meta inte
                d.Set("private_dns_hostname_type_on_launch", nil)
        }

-       if err := d.Set(names.AttrTags, keyValueTags(ctx, subnet.Tags).IgnoreAWS().IgnoreConfig(ignoreTagsConfig).Map()); err != nil {
-               return sdkdiag.AppendErrorf(diags, "setting tags: %s", err)
-       }
-
        d.Set(names.AttrVPCID, subnet.VpcId)

        return diags
```

This removes much of the manual manipulation of the `subnet.Tags` field, instead relying on our centralized tagging logic to handle reconciling and writing the final result. However, the switch to transparent tagging means an additional call to `DescribeTags` is now made (and the corresponding IAM permissions are now required). In this case, the extra API call is actually redundant - `Tags` was already returned in the `DescribeSubnets` API response - so we can alter the transparent tagging configuration to skip calling the subsequent `DescribeTags` API call. 

I _believe_ this should fix the issue, but I'll need to test more thoroughly tomorrow to make sure there wasn't a reason this approach was taken instead. Thanks for your patience and hopefully this helps to describe what happened here more completely in the interim.

--- 

As an aside - We don't currently make any compatibility promises around the IAM permissions required across minor versions. This is partly because changes like this are uncommon, but also because we don't have a holistic source of the permissions required for a given API which we could translate into the permissions required to operate a resource. Our approach to handling breaking IAM permission changes is something we're planning to align on and document more explicitly in the near future.
